### PR TITLE
Add support for referencable labels for attribute documentation

### PR DESCRIPTION
**Pull request by ojhunt**
The existing mechanism being used is to manually add a reference in the documentation. These references link to the beginning of the text rather than the heading for the attribute which is what this PR allows.

### CODE REVIEW

File: clang/utils/TableGen/ClangAttrEmitter.cpp
Comment (Sirraide) [inline clang/utils/TableGen/ClangAttrEmitter.cpp:5181]:
```suggestion
  if (StringRef label = Doc.Documentation->getValueAsString("Label");
```

**Review by Sirraide — COMMENTED** [review]
Implementation seems fine to me; I can’t speak as to whether this is a necessary addition though.

**Comment by erichkeane** [comment]
I have no idea what is going on here... @AaronBallman probably needs to take a better look.  Else, could the author please give me an ELI5 sorta thing here or show me what the change looks like on the generated stuff?

File: clang/include/clang/Basic/Attr.td
Comment (AaronBallman) [inline clang/include/clang/Basic/Attr.td:59]:
```suggestion
  // When set, specifies a label that can be used to reference the documentation.
```

File: clang/utils/TableGen/ClangAttrEmitter.cpp
Comment (AaronBallman) [reply clang/utils/TableGen/ClangAttrEmitter.cpp:5181]: Also: `label` -> `Label` per our usual naming rules.

**Review by AaronBallman — COMMENTED** [review]
Out of curiosity, why does the existing mechanism not suffice?

**Comment by ojhunt** [comment]
> Out of curiosity, why does the existing mechanism not suffice?

it "works" but creates the tag at the beginning of the text rather than the beginning of the attribute documentation. The result is that the scroll target for linking is the beginning of the text rather than the documentation as well, e.g. the location pointed to here:
Tag+Scroll Location

As a result when you follow the link the actual attribute name, syntax table, etc is not visible and it is necessary to scroll to see the actual documentation being linked.

**Comment by ojhunt** [comment]
> I have no idea what is going on here... @AaronBallman probably needs to take a better look. Else, could the author please give me an ELI5 sorta thing here or show me what the change looks like on the generated stuff?

Sure! This example will cover the docs used in the [above screenshot](#issuecomment-2516159888)

So currently the documentation for the attribute is:

```markdown
def NoSanitizeMemoryDocs : Documentation {
  let Category = DocCatFunction;
  let Heading = "no_sanitize_memory";
  let Content = [{
.. _langext-memory_sanitizer:

Use ``__attribute__((no_sanitize_memory))`` on a function declaration to
....
  }];
}
```

this results in generated docs like this:

```rst
no_sanitize_memory
------------------
.. csv-table:: Supported Syntaxes
   :header: "GNU", "C++11", "C23", "``__declspec``", "Keyword", "``#pragma``", "HLSL Annotation", "``#pragma clang attribute``"

   "``no_address_safety_analysis`` |br| ``no_sanitize_address`` |br| ``no_sanitize_thread`` |br| ``no_sanitize_memory``","``gnu::no_address_safety_analysis`` |br| ``gnu::no_sanitize_address`` |br| ``gnu::no_sanitize_thread`` |br| ``clang::no_sanitize_memory``","``gnu::no_address_safety_analysis`` |br| ``gnu::no_sanitize_address`` |br| ``gnu::no_sanitize_thread`` |br| ``clang::no_sanitize_memory``","","","","","Yes"

.. _langext-memory_sanitizer: <!-- NOTE: This is where the manually specified label is -->

Use ``__attribute__((no_sanitize_memory))`` on a function declaration to
...
```

The result is that a `:ref:langext-memory_sanitizer` link goes into the middle of the attribute documentation, after the name and spelling table.

This change simply adds a `Label` field to the root Documentation object and updates the codgen to use it, and removes the currently manually specified labels, after which we can remove the manual label from the doc string above and use the explicit Label field which then results in this generated documentation:

```rst
.. _langext-memory_sanitizer: <!-- NOTE: This is where the manually specified label is now -->

no_sanitize_memory
------------------
.. csv-table:: Supported Syntaxes
   :header: "GNU", "C++11", "C23", "``__declspec``", "Keyword", "``#pragma``", "HLSL Annotation", "``#pragma clang attribute``"

   "``no_address_safety_analysis`` |br| ``no_sanitize_address`` |br| ``no_sanitize_thread`` |br| ``no_sanitize_memory``","``gnu::no_address_safety_analysis`` |br| ``gnu::no_sanitize_address`` |br| ``gnu::no_sanitize_thread`` |br| ``clang::no_sanitize_memory``","``gnu::no_address_safety_analysis`` |br| ``gnu::no_sanitize_address`` |br| ``gnu::no_sanitize_thread`` |br| ``clang::no_sanitize_memory``","","","","","Yes"

Use ``__attribute__((no_sanitize_memory))`` on a function declaration to
...
```

This puts the link in a much better location for cross linking documentation

**Comment by Sirraide** [comment]
> This has no impact on the visual appearance of the resultant html documentation, but places the anchor in a much better location for cross linking documentation.

I see, I did’t quite pick up on that part because I don’t know RST very well, but I feel like that’s a reasonable addition then.

File: clang/utils/TableGen/ClangAttrEmitter.cpp
Comment (Sirraide) [inline clang/utils/TableGen/ClangAttrEmitter.cpp:5181]:
```suggestion
  if (StringRef Label = Doc.Documentation->getValueAsString("Label");
```
I think you may have missed this

**Review by Sirraide — APPROVED** [review]
LGTM after fixing one more thing.

I personally think this is a reasonable addition now, but please wait for Aaron or Erich to approve this too before merging.

**Comment by Sirraide** [comment]
Also, if/when this is merged, it would be nice if someone could go through the rest of the docs and apply this wherever possible (assuming there are any other places where we still need to do that).

**Comment by ojhunt** [comment]
> Also, if/when this is merged, it would be nice if someone could go through the rest of the docs and apply this wherever possible (assuming there are any other places where we still need to do that).

yeah, I wanted to isolate this for now to just the attribute docs as these are the only cases I could find, and I found them while working on the docs for #117428.

I'm kind of surprised I couldn't find other labels anywhere, but it could just be because other docs don't reference each other due to the inability to easily do so?

**Comment by ojhunt** [comment]
> LGTM after fixing one more thing.
> 
> I personally think this is a reasonable addition now, but please wait for Aaron or Erich to approve this too before merging.

Righto, I've committed the change, is `const StringRef` a style violation (I recognize StringRef is a constant type)? (Pondering adding support for style rule enforcement for things like "this is a constant type so doesn't need to be const")

**Comment by Sirraide** [comment]
> is `const StringRef` a style violation

Yeah, we generally don’t use top-level `const` for variables that are value types.

**Review by AaronBallman — APPROVED** [review]
Ah thank you for the more detailed explanation, this makes sense to me. LG!

**Comment by erichkeane** [comment]
Yeah, agreed, thanks for the ELI5.  When it comes to RST/HTML, I'm roughly 5 :)  LGTM as well.

**Pull request merged**
